File: Arithmetic Subarrays.py
class Solution:
    def checkArithmeticSubarrays(self, nums: List[int], l: List[int], r: List[int]) -> List[bool]:
        res = list()
        for i in range(len(l)):
            sub_array = nums[l[i]:r[i]+1]
            sub_array.sort()
            flag = True
            for j in range(len(sub_array)-1):
                if sub_array[0] - sub_array[1] != sub_array[j] - sub_array[j+1]:
                    flag = False
                    break
            if flag is False:
                res.append(False)
            else:
                res.append(True)
        return res

File: POD/Week 1 - Arrays/Arithmetic_Subarrays.cpp
class Solution
{
public:
    vector<bool> checkArithmeticSubarrays(vector<int> &nums, vector<int> &l, vector<int> &r)
    {
        //vector to store answers
        vector<bool> ans;

        //for each subarray - O(m) where m is the length of l and r
        for (int i = 0; i < l.size(); i++)
        {
            //get indices to find subarray
            auto start = nums.begin() + l[i];
            auto end = nums.begin() + r[i] + 1;
            //declaring subarray
            vector<int> subarray(r[i] - l[i] + 1); 
            copy(start, end, subarray.begin());

            //sort subarray - O(xlogx) where x is the size of subarray
            sort(subarray.begin(), subarray.end());

            bool flag = true;

            //check if subarray is arithmetic - O(x) where x is the size of subarray
            for (int j = 0; j < subarray.size() - 1; j++)
            {
                //if difference between adjacent elements is not 1
                if (subarray[j + 1] - subarray[j] != subarray[1] - subarray[0])
                {
                    flag = false;
                    break;
                }
            }
            //push answer to vector
            ans.push_back(flag);
        }
        return ans;
    }
};

// time complexity: O(mxlogx)
// space complexity: O(mx) for answer and subarray vectors